File: api-aggregator/src/test/java/com/github/alexeysol/geekregime/apiaggregator/features/user/service/v1/userservice/FindUserByIdTest.java
package com.github.alexeysol.geekregime.apiaggregator.features.user.service.v1.userservice;

import com.github.alexeysol.geekregime.apicommons.exception.SerializedApiException;
import com.github.alexeysol.geekregime.apicommons.model.dto.user.UserDto;
import com.github.tomakehurst.wiremock.client.ResponseDefinitionBuilder;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;

import static com.github.tomakehurst.wiremock.client.WireMock.*;

public class FindUserByIdTest extends BaseUserServiceTest {
    @Test
    public void userExists_whenFindUserById_thenReturnsUser() {
        long userId = 1L;
        String path = String.format("%s/%d", API_USERS_PATH, userId);

        ResponseDefinitionBuilder responseToReturn = aResponse()
            .withStatus(HttpStatus.OK.value())
            .withBodyFile(getJsonPath("getUser", HttpStatus.OK));

        wireMockServer.stubFor(getApiMappingBuilder(path, responseToReturn));

        UserDto user = service.findUserById(userId);

        wireMockServer.verify(getRequestedFor(urlPathEqualTo(path))
            .withHeader("Content-Type", equalTo("application/json")));

        Assertions.assertNotNull(user);
        Assertions.assertEquals(userId, user.getId());
    }

    @Test
    public void userDoesntExist_whenFindUserById_thenReturnsStatus404() {
        long absentId = 10L;
        String path = String.format("%s/%d", API_USERS_PATH, absentId);

        ResponseDefinitionBuilder responseToReturn = aResponse()
            .withStatus(HttpStatus.NOT_FOUND.value())
            .withBodyFile(getJsonPath("getUser", HttpStatus.NOT_FOUND));

        wireMockServer.stubFor(getApiMappingBuilder(path, responseToReturn));

        Assertions.assertThrows(
            SerializedApiException.class, () -> service.findUserById(absentId)
        );

        wireMockServer.verify(getRequestedFor(urlPathEqualTo(path))
            .withHeader("Content-Type", equalTo("application/json")));
    }
}

File: api-users/src/test/kotlin/com/github/alexeysol/geekregime/apiusers/service/v1/userservice/FindUserByIdTest.kt
package com.github.alexeysol.geekregime.apiusers.service.v1.userservice

import com.github.alexeysol.geekregime.apiusers.model.entity.User
import io.mockk.every
import io.mockk.verify
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions

class FindUserByIdTest : BaseUserServiceTest() {
    @Test
    fun givenUserExists_whenFindUserById_thenReturnsUser() {
        val userId = 1L
        val user = User(email = "mark@mail.com", details = defaultDetails)

        every { repository.findUserById(userId) } returns user

        val result = service.findUserById(userId)
        verify(exactly = 1) { repository.findUserById(userId) }
        Assertions.assertEquals(user, result)
    }

    @Test
    fun givenUserDoesntExist_whenFindUserById_thenReturnsNull() {
        val absentUserId = 10L

        every { repository.findUserById(absentUserId) } returns null

        val result = service.findUserById(absentUserId)
        verify(exactly = 1) { repository.findUserById(absentUserId) }
        Assertions.assertEquals(null, result)
    }
}
